python/src/aac/lang/references.py

```diff
@@ -21,3 +22,198 @@ def filter_definitions_by_reference(definition_to_filter: Definition) -> bool:
+
+    return keepers
+
+
+def _process_segment(prefix: list[str], segments: list[str], definitions: list[Definition], language_context: LanguageContext) -> list[Definition]:
+    """This should process segments recursively."""
+    if len(segments) == 0:
+        # everything has been processed, so return what was found in the previous recursion
+        return definitions
+
+    keepers = []
+    key, selector, selector_field, selector_value = _get_reference_segment_content(segments[0])
+
+    if key in language_context.get_root_keys():
+        # handle root key
+        keepers.extend(_process_root(selector, selector_field, selector_value, definitions))
+
+    else:
+        # handle non-root keys
+        keepers.extend(_process_non_root(key, prefix, selector, selector_field, selector_value, definitions))
+
+    if len(keepers) == 0:
+        return []
+
+    nxt_prefix = prefix
+    nxt_prefix.append(key)
+    return _process_segment(nxt_prefix, segments[1:], keepers, language_context)
+
+
+def _process_root(selector: str, selector_field: str, selector_value: str, definitions: list[Definition]) -> list[Definition]:
+    """Return definitions that match the root selector.  If no selector, return all definitions."""
+    keepers = []
+    if selector:
+        # handle root level selector
+        for definition in definitions:
+            if str(definition.get_top_level_fields()[selector_field]) == str(selector_value):
+                keepers.append(definition)
+    else:
+        # no root level selector, so just keep all definitions and move on
+        keepers.extend(definitions)
+    return keepers
+
+
+def _process_non_root(key: str, prefix: list[str], selector: str, selector_field: str, selector_value: str, definitions: list[Definition]) -> list[Definition]:
+    # initialize return value as empty string
+    keepers = []
+    for definition in definitions:
+        top_level_fields = definition.get_top_level_fields()
+        dict_list = _drill_into_nested_dict(prefix[1:], top_level_fields)
+
+        for def_dict in dict_list:
+            if key in def_dict.keys():
+                if selector:
+                    # since there is a selector, only keep the definition if it contains the specified selector field and value
+                    # note that selectors filter based on child field values
+                    value_for_key = def_dict[key]
+                    if type(value_for_key) is list:
+                        for item in value_for_key:
+                            if str(item[selector_field]) == str(selector_value):  # casting these seems like a hack
+
+                                keepers.append(definition)
+                                break
+                    else:
+                        if str(value_for_key[selector_field]) == str(selector_value):  # casting these seems like a hack
+                            keepers.append(definition)
```

Comment:
It looks like you should be able to remove the string casting -- the tests still pass after I removed them. 

I'm happy to help identify/test a scenario in which we're expecting non-string values in the comparisons.
```suggestion
                    if isinstance(value_for_key, list):
                        for item in value_for_key:
                            if item.get(selector_field) == selector_value:
                                keepers.append(definition)
                                break
                    elif value_for_key.get(selector_field) == selector_value:
                        keepers.append(definition)
```

Reply:
I put that cast in because I ran into some hard to diagnose python errors while I was doing development.  It may have become a non-issue as I worked out some bugs, but I'm still hesitant to make to make the change.  As a compromise, I'll add a new issue to expand the test cases and remove casts if they all pass.  I think the issue may have occurred when the field was an integer.  Now I realize I didn't explicitly create a unit test for that condition.